protocols/gossipsub/src/handler.rs

```diff
@@ -564,17 +495,58 @@ impl ConnectionHandler for Handler {
             Self::OutboundOpenInfo,
         >,
     ) {
+        if event.is_outbound() {
+            self.outbound_substream_establishing = false;
+        }
+
+        if event.is_inbound() && self.inbound_substreams_created == MAX_SUBSTREAM_CREATION {
+            // Too many inbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::warn!("The maximum number of inbound substreams created has been exceeded");
+            return;
+        }
+
+        if event.is_outbound() && self.outbound_substreams_requested == MAX_SUBSTREAM_CREATION {
+            // Too many outbound substreams have been created, disable the handler.
+            self.keep_alive = KeepAlive::No;
+            log::warn!("The maximum number of outbound substreams created has been exceeded");
+            return;
+        }
```

Comment:
As far as I can tell there are still multiple issues with this code.

- In case the remote node joins the local nodes mesh (`HandlerIn::JoinedMesh`), `self.keep_alive = KeepAlive::No` will be replaced by `KeepAlive::Yes`, see `on_behaviour_event`. Also in case the remote node leaves the local nodes mesh, `KeepAlive::Until` is set. Though the `ConnectionHandler` will still not be able to establish an outbound stream as `MAX_SUBSTREAM_CREATION` is hit.
- We may still emit `OutboundSubstreamRequest` in `ConnectionHandler::poll` as it does not check `self.outbound_substreams_requested <= MAX_SUBSTREAM_CREATION` first. That leads to wasted stream upgrades as that stream will then be dropped here.

I will have to think about this with a clear mind, i.e. tomorrow. Maybe we should go @thomaseizinger suggested route of modelling all of these states (keep alive, disabling, not supported, ...) as a combined `enum` to reduce the amount of possible state variants at compile time.

Reply:
See https://github.com/libp2p/rust-libp2p/pull/3625#issuecomment-1498924751 for a potential solution.